spec/abstract/KHeap_A.thy

```diff
@@ -624,4 +624,44 @@ definition
 throw_on_false :: "'e \<Rightarrow> (bool,'z::state_ext) s_monad \<Rightarrow> ('e + unit,'z::state_ext) s_monad" where
 "throw_on_false ex f \<equiv> doE v \<leftarrow> liftE f; unlessE v $ throwError ex odE"
 
+text \<open>Allow preemption at this point.\<close>
+
+definition
+  is_cur_domain_expired :: "'z::state_ext state \<Rightarrow> bool"
+where
+  "is_cur_domain_expired = (\<lambda>s. domain_time s < consumed_time s + MIN_BUDGET)"
+
+text \<open>Update current and consumed time.\<close>
+definition
+  update_time_stamp :: "(unit, 'z::state_ext) s_monad"
+where
+  "update_time_stamp = do
+    prev_time \<leftarrow> gets cur_time;
+    cur_time' \<leftarrow> do_machine_op getCurrentTime;
+    modify (\<lambda>s. s\<lparr> cur_time := cur_time' \<rparr>);
+    modify (\<lambda>s. s\<lparr> consumed_time := consumed_time s + cur_time' - prev_time \<rparr>)
+  od"
+
+definition
+  preemption_point :: "(unit,'z::state_ext) p_monad" where
+ "preemption_point \<equiv> doE
+    liftE $ do_extended_op update_work_units;
+    OR_choiceE work_units_limit_reached
+    (doE liftE $ do_extended_op reset_work_units;
+         irq_opt \<leftarrow> liftE $ do_machine_op (getActiveIRQ True);
+         case irq_opt of None \<Rightarrow> do update_time_stamp;
```

Comment:
If you make my suggested change to the Haskell then you would want to also change this `case` to an `if`.

Reply:
Thanks Corey. I'll look into this!

Reply:
How does the change to `mcsIRQ` in `callKernel` relate to `preemptionPoint/preemption_point`?

Reply:
Oh sorry, I just realised you were talking about your other comment. I like your idea. Thanks!

Reply:
I thought that I could write `if (bound irq_opt)`, but `bound` is defined only in `InvariantsPre_AI.thy`. Should I move this to `KHeap_A.thy` or `lib` somewhere? I've currently just got `if (\exists y. irq_opt = Some y)`

Reply:
`bound` is actually an abbreviation of `blah ~= None`, which I don't like all that much for this particular use case. So I'm just going to leave it as is for now